packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,99 @@
+const ERROR_MESSAGE_NUMBER = (value) => `${ERROR_MESSAGE_PREFIX} number "${value}".`;
+const ERROR_MESSAGE_ADDRESS = (value) => `${ERROR_MESSAGE_PREFIX} address "${value}".`;
+const ERROR_MESSAGE_CHECKSUM_ADDRESS = (value) => `${ERROR_MESSAGE_PREFIX} address "${value}" checksum fails.`;
+const ERROR_MESSAGE_BYTES = (value) => `${ERROR_MESSAGE_PREFIX} bytes "${value}".`;
+const ERROR_MESSAGE_STRING = (value) => `${ERROR_MESSAGE_PREFIX} string "${value}".`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], rawValues: any[] = []): string {
+  if(types.length !== rawValues.length) throw new Error('Supplied number of types and values do not match.');
+  const values = [];
+  _.zipWith(types, rawValues, function(type: string, value: any) {
+    values.push(parseTypeValuePair(type, value));
+  });
+  const methodId = abi.methodID(name, types).toString('hex');
+  const params = abi.rawEncode(types, values).toString('hex');
   return '0x' + methodId + params;
 }
+
+export function parseTypeValuePair(type: string, rawValue: any): string | never {
+  if(type === 'address') {
+    return parseAddress(rawValue);
+  }
+  else if(type === 'string') {
+    return rawValue; // Validated by ethereumjs-abi.
+  }
+  else if(type.startsWith('bytes') && !type.includes('[]')) {
+    return parseBytes(rawValue);
+  }
+  else if(type.startsWith('uint') && !type.includes('[]')) {
+    return parseNumber(rawValue, true);
+  }
+  else if(type.startsWith('int') && !type.includes('[]')) {
+    return parseNumber(rawValue, false);
+  }
+  else {
+    // TODO: parse remianing types: fixed, arrays, etc.
+    return rawValue;
+  }
+}
+
+function parseBytes(value: any): string | never {
+  if(typeof(value) !== 'string') throw new Error(ERROR_MESSAGE_BYTES(value));
+  if(value.toString().length === 0) return value;
+  if(!/^(0x)?[0-9a-f]$/i.test(value)) throw new Error(ERROR_MESSAGE_BYTES(value));
+  return value;
+}
+
+function parseAddress(value: any): string | never {
+  if(typeof(value) !== 'string') throw new Error(ERROR_MESSAGE_ADDRESS(value));
+  if(!/^(0x)?[0-9a-f]{40}$/i.test(value)) throw new Error(ERROR_MESSAGE_ADDRESS(value));
+  // TODO: It'd be nice to perform a checksum validation here.
+  return value;
+}
+
+function parseNumber(value: any, mustBePositive: boolean): string | never {
+  if(typeof(value) === 'number') return parseNumberFromLiteral(value, mustBePositive);
+  else if(typeof(value) === 'string') return parseNumberFromString(value, mustBePositive);
+  else if(BN.isBigNumber(value)) return parseNumberFromBigNumber(new BN(value), mustBePositive);
+  throw new Error(ERROR_MESSAGE_NUMBER(value));
```

Comment:
Question about ts: is it possible to dispatch or pattern match based on the type of `value` in runtime? I mean, is it possible to have the functions `parseNumberFromLiteral`, `parseNumberFromString` and `parseNumberFromBigNumber` changed to just `parse(value: BN)`, `parse(value: string)`, and `parse(value: number)`, and then ts chooses which function to call?

Reply:
I'm afraid not. Or at least I can't think of any...
The closes thing I can think of is:
```function parseNumber(rawValue: number, mustBePositive: boolean): string | never;
function parseNumber(rawValue: string, mustBePositive: boolean): string | never;
function parseNumber(rawValue: BN, mustBePositive: boolean): string | never;
function parseNumber(rawValue: number | string | BN, mustBePositive: boolean): string | never {
  // Code...
  return 'a';
}
```
In Ts, you have to write method declarations for each of the overloads and then one method that takes a union of all the options that contains the actual code. Simpler overloading is supported when the number of parameters is different. I guess this is because if everything get's translated to Js, you have to perform the type checks anyway.
More info here: https://stackoverflow.com/questions/12688275/is-there-a-way-to-do-method-overloading-in-typescript/12689054#12689054